File: app_desktop/models/game_model.py
from models.model import Model
from datetime import datetime, date


class GameModel(Model):

    def find_all_games(self):
        db = self.db
        today = str(date.today())
        cursor = db.cursor(dictionary=True)
        cursor.execute(("SELECT game.id_game, t1.name AS t1_name, t2.name AS t2_name, "
                        "DATE_FORMAT(start_time, '%Hh%i') AS start_date, "
                        "DATE_FORMAT(end_time, '%Hh%i') AS end_date "
                        "FROM game "
                        "JOIN team t1 ON game.id_team1 = t1.id_team "
                        "JOIN team t2 ON game.id_team2 = t2.id_team WHERE DATE(start_time) = %s"), (today,))
        games = cursor.fetchall()
        cursor.close()
        db.close()
        return games

    def find_one_game(self, id_game):
        db = self.db
        cursor = db.cursor(dictionary=True)
        cursor.execute(("SELECT *, game.id_game, t1.name AS t1_name, t2.name AS t2_name, "
                        "DATE_FORMAT(start_time, '%Hh%i') AS start_date, "
                        "DATE_FORMAT(end_time, '%Hh%i') AS end_date "
                        "FROM game "
                        "JOIN team t1 ON game.id_team1 = t1.id_team "
                        "JOIN team t2 ON game.id_team2 = t2.id_team WHERE id_game = %s"), (id_game,))
        game = cursor.fetchone()
        cursor.close()
        return game

    def find_bets_nb(self, id_game, id_team):
        db = self.db
        cursor = db.cursor(dictionary=True)
        cursor.execute("SELECT COUNT(id_bet) as bets_nb FROM bet WHERE id_game = %s AND id_team = %s", (id_game, id_team))
        nb = cursor.fetchone()
        nb = nb['bets_nb']
        cursor.close()
        return nb

    def find_comments(self, id_game):
        db = self.db
        cursor = db.cursor(dictionary=True)
        cursor.execute("SELECT *  FROM comment WHERE id_game = %s ", (id_game,))
        comments = cursor.fetchall()
        cursor.close()
        return comments

    def close_game(self, id_game):
        db = self.db
        today = str(datetime.today())
        cursor = db.cursor(dictionary=True)
        cursor.execute("UPDATE game SET end_time = %s WHERE id_game = %s", (today, id_game))
        cursor.execute("SELECT * FROM game WHERE id_game = %s", (id_game,))
        game = cursor.fetchone()
        cursor.execute("SELECT * FROM bet WHERE id_game = %s", (id_game,))
        bets = cursor.fetchall()
        for bet in bets:
            if game['score_1'] > game['score_2']:
                id_team = game['id_team1']
                odds = game['odds_1']
            else:
                id_team = game['id_team2']
                odds = game['odds_2']
            if bet['id_team'] == id_team:
                cursor.execute("UPDATE bet SET profits = %s WHERE id_bet = %s", (bet['amount'] * odds, bet['id_bet']))
            else:
                cursor.execute("UPDATE bet SET losses = %s WHERE id_bet = %s", (0 - bet['amount'], bet['id_bet']))
        db.commit()

File: app_web/app/models/GameModel.php
<?php

class GameModel extends Model{
    
    public function findAllGamesToday(): array
    {
        try {
            $query = "SELECT game.id_game, t1.name AS t1_name, t2.name AS t2_name, DATE_FORMAT(start_time, '%Hh%i') AS game_date
            FROM game
            JOIN team t1 ON game.id_team1 = t1.id_team
            JOIN team t2 ON game.id_team2 = t2.id_team
            WHERE DATE(start_time) = CURDATE();";
            $query = $this->pdo->query($query);
            return $query->fetchAll();
        } catch (PDOException $dbError) {
            throw new Exception("Erreur de la base de données.");
        }
    }
    
    public function findOneGameById(int $id): array
    {
        try {
            $query = $this->pdo->prepare("SELECT *, t1.name AS t1_name, t2.name AS t2_name, DATE_FORMAT(end_time, '%Hh%i') AS end, DATE_FORMAT(start_time, '%d/%m/%Y - %Hh%i') AS game_date
            FROM game
            JOIN team t1 ON game.id_team1 = t1.id_team
            JOIN team t2 ON game.id_team2 = t2.id_team
            WHERE game.id_game = :id;");
            $query->execute(['id' => $id]);
            $game = $query->fetch();
            $game['status'] = $this->gameStatus($game['start_time'], $game['end_time']);
            return $game;
        } catch (PDOException $e) {
            throw new Exception("Erreur de la base de données.");
        }
    }

    public function findAllGames(): array
    {
        try {
            $query = "SELECT *, t1.name AS t1_name, t2.name AS t2_name, DATE_FORMAT(start_time, '%d/%m/%Y - %Hh%i') AS game_date
            FROM game
            JOIN team t1 ON game.id_team1 = t1.id_team
            JOIN team t2 ON game.id_team2 = t2.id_team";
            $query = $this->pdo->query($query);
            $games = $query->fetchAll();
            foreach ($games as $i => $game) {
                $games[$i]['status'] = $this->gameStatus($game['start_time'], $game['end_time']);
            }
            return $games;
        } catch (PDOException $e) {
            throw new Exception("Erreur de la base de données.");
        }
    }
    
    public function findAllGamesById(array $id_games): array
    {
        try {
            $query = "SELECT *, t1.name AS t1_name, t2.name AS t2_name, DATE_FORMAT(start_time, '%d/%m/%Y - %Hh%i') AS game_date
            FROM game
            JOIN team t1 ON game.id_team1 = t1.id_team
            JOIN team t2 ON game.id_team2 = t2.id_team
            WHERE id_game IN (".implode(',', $id_games).")";
            $query = $this->pdo->query($query);
            return $query->fetchAll();
        } catch (PDOException $e) {
            throw new Exception("Erreur de la base de données.");
        }
    }

    // Détermine le statut d'un match
    public static function gameStatus($start_time, $end_time): string
    {
        if($end_time) {
            return 'Terminé';
        } elseif(date("Y-m-d H:i:s") > $start_time){
            return 'En cours';
        } else {
            return 'À venir';
        } 
    }
    
    // Ajout d'un match
    public function addGame(int $id_team1, float $odds_1, int $id_team2, float $odds_2, string $date) : void
    {   
        try {
            $query = $this->pdo->prepare("INSERT INTO game (id_team1, odds_1, id_team2, odds_2, start_time) VALUES (:id_team1, :odds_1, :id_team2, :odds_2, :date);");
            $query->execute([
                ':id_team1' => $id_team1,
                ':odds_1' => $odds_1,
                ':id_team2' => $id_team2,
                ':odds_2' => $odds_2,
                ':date' => $date
            ]);
        } catch (PDOException $e) {
            throw new Exception("Erreur de la base de données.");   
        }   
    } 
}